File: typescript-automation/src/dsl/protocols/driverRegistry.ts
import type { Channel } from "./channel";
import type { CatalogueProtocol } from "./catalogueProtocol";

type DriverFactory = () => CatalogueProtocol;

const registry = new Map<Channel, DriverFactory>();

export function register(channel: Channel, factory: DriverFactory): void {
  registry.set(channel, factory);
}

export function create(channel: Channel): CatalogueProtocol {
  const factory = registry.get(channel);
  if (factory === undefined) {
    throw new Error(
      `No driver registered for channel '${channel}'. Valid values: ${[...registry.keys()].join(", ")}`
    );
  }
  return factory();
}

File: python-automation/src/automation/dsl/protocols/driver_registry.py
import importlib
from collections.abc import Callable

from automation.dsl.protocols.catalogue_protocol import CatalogueProtocol
from automation.dsl.protocols.channel import Channel

_registry: dict[Channel, Callable[[], CatalogueProtocol]] = {}


def register(channel: Channel, factory: Callable[[], CatalogueProtocol]) -> None:
    _registry[channel] = factory


def create(channel: Channel) -> CatalogueProtocol:
    factory = _registry.get(channel)
    if factory is None:
        _load_by_convention(channel)
        factory = _registry.get(channel)
    if factory is None:
        raise RuntimeError(
            f"No driver registered for channel '{channel.value}'. "
            f"Registered: {[c.value for c in _registry.keys()]}"
        )
    return factory()


def _load_by_convention(channel: Channel) -> None:
    module_name = f"automation.driver.{channel.name.lower()}.my_ecommerce_driver"
    try:
        importlib.import_module(module_name)
    except ImportError:
        pass
